File: AlgorithmsOnGraphs-UCSD/week3/bipartite.py
#Uses python3

import sys

#m coloring problem where m = 2
def bipartite(adj):
    #write your code here
    colors = len(adj) * [-1]
    queue = []
    colors[0] = 1
    queue.append(0)
    
    while not queue:
        u = queue.pop(0)
        #if adj[u][u] == 1: #self loop
        #    return 0
        
        for v in adj[u]:
            #destination not colored
            if colors[v] == -1:
                colors[v] = 1 - colors[u]
                queue.append(v)

            elif colors[v] == colors[u]: #adjacent nodes having same color
                return 0
    return 1

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    print(bipartite(adj))

File: AlgorithmsOnGraphs-UCSD/week3/bipartite.cpp
//graph m coloring problem --> find the bipartiteness
#include <iostream>
#include <vector>
#include <queue>

using std::vector;
using std::queue;

int bipartite(vector<vector<int> > &adj) {
    //write your code here
    // we create a color array to store colors assigned to all veritces.
    // value 1 --> color 1 assigned
    // value 0 --> color 2 assigned
    vector<int> colorArr(adj.size(), -1);

    //assign color to source vertex
    colorArr[0] = 1;

    //enqueue source
    queue<int> queue;
    queue.push(0);

    while(!queue.empty()){
        int u = queue.front();
        queue.pop();

        //check for self loop
        //if(adj[u][u] == 1) 
        //    return 0;
        
        //find all non-colored adjacent vertices
        for(auto it = adj[u].begin(); it != adj[u].end(); it++) {
            int v = *it;

            //destination v is not colored
            if (colorArr[v] == -1) {
            //assign alternate color to this adjacent v of u
                colorArr[v] = 1- colorArr[u];
                queue.push(v);
            }
            
            //destination v is colored with same color as u
            else if (colorArr[v] == colorArr[u])
                return 0;
        }
    }
    // If we reach here, then all adjacent vertices can
    //  be colored with alternate color
    return 1;
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
    adj[y - 1].push_back(x - 1);
  }
  std::cout << bipartite(adj);
}
